modules/core/src/test/java/org/apache/ignite/internal/processors/security/cluster/ClusterStatePermissionTest.java

```diff
@@ -0,0 +1,344 @@
+        Initiator nodeType = initiator;
+        SecurityPermission[] perms = permissions;
+
+        initiator = Initiator.SERVER;
+        permissions = F.asArray(ADMIN_CLUSTER_STATE);
+
+        Ignite ig = startGrids(1);
+
+        ig.cluster().state(ACTIVE);
+
+        assertEquals(ACTIVE, ig.cluster().state());
+
+        initiator = nodeType;
+        permissions = perms;
+    }
+
+    /**
+     * Tries to change state and ensures that a proper error raises if required.
+     *
+     * @param states Cluster states to try to set to one by one.
+     * @param allowed {@code True}, if the state changes must be allowed. {@code False} otherwise.
+     */
+    private void doTestChangeState(ClusterState[] states, boolean allowed) throws Exception {
+        assert states != null && states.length > 0;
+
+        Ignite ig = startGrid(G.allGrids().size());
+
+        startGrid(G.allGrids().size());
+
+        Consumer<ClusterState> action = initiatorAction(ig);
+
+        for (int i = 0; i < states.length; ++i) {
+            ClusterState stateTo = states[i];
+
+            if (allowed) {
+                action.accept(stateTo);
+
+                assertEquals(stateTo, ig.cluster().state());
+            }
+            else
+                ensureThrows(action, stateTo);
+        }
+    }
+
+    /**
+     * Ensures that a proper error occurs on sluster change state action.
+     */
+    private void ensureThrows(Consumer<ClusterState> action, ClusterState stateTo) {
+        Class<? extends Throwable> cause = SecurityException.class;
+        String errMsg = "Authorization failed [perm=" + ADMIN_CLUSTER_STATE;
+
+        if (Initiator.THIN_CLIENT == initiator) {
+            cause = ClientAuthorizationException.class;
+            errMsg = "User is not authorized to perform this operation";
+        }
+        else if (Initiator.REMOTE_CONTROL == initiator)
+            cause = GridClientException.class;
+
+        assertThrowsAnyCause(
+            null,
+            () -> {
+                action.accept(stateTo);
+
+                return null;
+            },
+            cause,
+            errMsg
+        );
+    }
+
+    /**
+     * @return Change state operation depending on {@link #initiator}.
+     */
+    private Consumer<ClusterState> initiatorAction(Ignite srv) throws Exception {
+        switch (initiator) {
+            case SERVER:
+                return (state) -> srv.cluster().state(state);
+
+            case CLIENT:
+                return (state) -> {
+                    try {
+                        startClientGrid(G.allGrids().size()).cluster().state(state);
+                    }
+                    catch (Exception e) {
+                        throw new IgniteException("Unable to start client grid.", e);
+                    }
+                };
+
+            case THIN_CLIENT: {
+                return (state) -> G.startClient(new ClientConfiguration().setAddresses(Config.SERVER)
```

Comment:
IMHO thin clients should be stopped gracefully. Also take a look at the Grid Client.

Reply:
Fixed